tools/downloader/src/open_model_zoo/model_tools/download_engine/cache.py

```diff
@@ -0,0 +1,100 @@
+# Copyright (c) 2021 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import hashlib
+import shutil
+import sys
+import tempfile
+
+from pathlib import Path
+
+CHUNK_SIZE = 1 << 15 if sys.stdout.isatty() else 1 << 20
+
+
+class NullCache:
+    def has(self, hash): return False
+    def get(self, model_file, path, reporter): return False
+    def put(self, hash, path): pass
+
+
+class DirCache:
+    _FORMAT = 1 # increment if backwards-incompatible changes to the format are made
+    _HASH_LEN = hashlib.sha256().digest_size
+
+    def __init__(self, cache_dir):
+        self._cache_dir = cache_dir / str(self._FORMAT)
+        self._cache_dir.mkdir(parents=True, exist_ok=True)
+
+        self._staging_dir = self._cache_dir / 'staging'
+        self._staging_dir.mkdir(exist_ok=True)
+
+    def _hash_path(self, hash):
+        assert len(hash) == self._HASH_LEN
+        hash_str = hash.hex().lower()
+        return self._cache_dir / hash_str[:2] / hash_str[2:]
+
+    def has(self, hash):
+        return self._hash_path(hash).exists()
+
+    def get(self, model_file, path, reporter):
+        cache_path = self._hash_path(model_file.sha256)
+        cache_sha256 = hashlib.sha256()
+        cache_size = 0
+
+        with open(cache_path, 'rb') as cache_file, open(path, 'wb') as destination_file:
+            while True:
+                data = cache_file.read(CHUNK_SIZE)
+                if not data:
+                    break
+                cache_size += len(data)
+                if cache_size > model_file.size:
+                    reporter.log_error("Cached file is longer than expected ({} B), copying aborted", model_file.size)
+                    return False
+                cache_sha256.update(data)
```

Comment:
side comment: did we test if using shutil.copyfile and reading destination file for sha256 calculation is slower that doing both in one shot like in this implementation?